test/model/autocomplete_test.dart

```diff
@@ -274,37 +276,6 @@ void main() {
     }
   });
 
-  test('MentionAutocompleteView mutating store.users while in progress causes retry', () async {
```

Comment:
> The resulting test should
> 
> * pass with the code from main;
> * but fail with the code from main if you remove the retry.

I wonder if you meant:
* **fail** with the code from main; (because it still uses retry and this test is about not retrying)
* but **pass** with the code from main if you remove the retry.

Reply:
No, I think I had that the right way around.

The point is that this revised test is testing something *less specific* than what the existing test in main is testing:
* The existing test checks that the search completes and the search results reflect the newly-added user.
* The revised test checks that the search completes.

So both the code in main, and the code at the tip of the PR branch, pass the revised test because they make the search complete.

But if you start from main, and then take out the retry, it won't pass even the revised, less-specific test. That's because in main (where the search is iterating through the users map itself, and so is subject to ConcurrentModificationError), the retry is necessary for how it ensures the search completes.

Reply:
> The resulting test should
>
> * pass at the tip of your branch;
> * pass with the code from main;
> * but fail with the code from main if you remove the retry.

I tried to make this test as such to pass all of these conditions, but unfortunately, I wasn't able to do so. Here is the reason:

As you mentioned in [#693 comment](https://github.com/zulip/zulip-flutter/pull/693#discussion_r1631861197), to make sure `addUser` actually happened during the search, we should add `check(done).isFalse()` after `addUser` and this check will only pass if `store.users >= 2000`, because after the first batch of 1000 users is processed, we have one `check(done).isFalse()` check passed before `addUser` and to make sure `check(done).isFalse()` after `addUser` passes, we have to process another batch of 1000 users.
Now, this isn't the case with the code from main. The code from main does two things (retrying the search + doing the duplicate search through `refreshStaleUserResults`). The first batch of 1000 users is processed the same way as in this PR, means `check(done).isFalse()` before `addUser` is passed, then during the second batch, `addUser` is called that causes a `ConcurrentModificationError` which discards the second batch starting the search from start, one through `refreshStaleUserResults` and another through retrying. After the first batch of 1000 users is processed, `check(done).isFalse()` check after `addUser` is passed, then the second batch of 1000 users is processed moving to `check(done).isTrue()` check where it fails, because we have `store.users >= 2000` and it would pass if `store.users < 2000`, but then again it would not pass with the tip of this branch. So basically the main problem here that prevents this test from passing both in this branch and the main is that after the user is added, main is processing the users from the start and the current branch continues.
Apart from this, if we assume this problem doesn't exist (one possible solution: `store.users < 2000` and `await Future(() {}); check(done).isFalse();` before `addUser` is removed), then this test will fail for another reason: in the main we expect the final results to contain the added user, and in the current PR we don't expect it. 

I tried to come up with other variations of the test, but unfortunately, I wasn't able to make it pass for both conditions. Help from your side is strongly appreciated! 🙂

Reply:
> As you mentioned in [#693 comment](https://github.com/zulip/zulip-flutter/pull/693#discussion_r1631861197), to make sure `addUser` actually happened during the search, we should add `check(done).isFalse()` after `addUser` and this check will only pass if `store.users >= 2000`, because after the first batch of 1000 users is processed, we have one `check(done).isFalse()` check passed before `addUser` and to make sure `check(done).isFalse()` after `addUser` passes, we have to process another batch of 1000 users.

Yeah. The exact number of times one has to wait there doesn't really matter for the test's purposes — it's only important that it hasn't yet finished at the start, it eventually does finish, and that it doesn't take some huge number of iterations. So it's fine to say something like this:
```dart
    await store.addUser(eg.user(userId: 11000, email: 'user11000@example.com', fullName: 'User 11000'));
    await Future(() {});
    check(done).isFalse();
    for (int i = 0; i < 3; i++) {
      await Future(() {});
      if (done) break;
    }
    check(done).isTrue();
```



> this test will fail for another reason: in the main we expect the final results to contain the added user, and in the current PR we don't expect it.

Hmm, true. I guess what I should have said, then, is: if you delete that check (so that it doesn't say anything about whether the new user is in the results or not), then it meets those three conditions of pass/pass/fail. So deleting that check goes along with `git checkout origin -- lib` as part of trying the test on the other versions.

(But then the actual version of the test that gets committed should still have that check; if the code changes so that it starts including the result after all, we'll want to know.)